<a href="https://colab.research.google.com/github/Konstantin-Solomin/Yandex-Practicum-Data-Science/blob/main/bank.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Исследование надежности заемщиков


In [ ]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

In [ ]:
data['days_employed'] = data['days_employed'].abs()

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

In [ ]:
data['education'] = data['education'].str.lower()

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [ ]:
data.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose', 'total_income_category', 'purpose_category'],
      dtype='object')

In [ ]:
data.groupby('children')['children'].count() 

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

In [ ]:
data.groupby('children')['debt'].sum() 

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

In [ ]:
data.groupby('children')['debt'].mean().sort_values().tail(5)

children
0    7.543822
3    8.181818
1    9.234609
2    9.454191
4    9.756098
dtype: float64

In [ ]:
t = (data.groupby('children')['debt'].sum() / data.groupby('children')['children'].count() * 100).sort_values().tail(5)

t

children
0    7.543822
3    8.181818
1    9.234609
2    9.454191
4    9.756098
dtype: float64

In [ ]:
tmp.values[1:].sum() / 4

9.156678940239289

**Вывод:** Данные указывают на слабую зависимость между количеством детей и вероятностью возникновения задолженности у заемщика. Следует также учитывать значительную разницу в размерах оцениваемых групп. Таким образом следует признать, что количество детей не влияет на вероятность возникновения задолженности у заемщика. Но определенно влияет отсутствие детей, так как в исследуемой группе без детей, случаи задолженности возникали значительно реже.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
data.groupby('family_status')['family_status'].count() 

family_status
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
гражданский брак          4134
женат / замужем          12261
Name: family_status, dtype: int64

In [ ]:
data.groupby('family_status')['debt'].sum() 

family_status
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
гражданский брак         385
женат / замужем          927
Name: debt, dtype: int64

In [ ]:
(data.groupby('family_status')['debt'].sum() / data.groupby('family_status')['family_status'].count() * 100).sort_values()

family_status
вдовец / вдова           6.624606
в разводе                7.064760
женат / замужем          7.560558
гражданский брак         9.313014
Не женат / не замужем    9.763948
dtype: float64

In [ ]:
data.groupby('family_status')['dob_years'].median().sort_values(ascending=False)

family_status
вдовец / вдова           58
в разводе                46
женат / замужем          43
гражданский брак         41
Не женат / не замужем    36
Name: dob_years, dtype: int64

**Вывод:** Мы обнаружили, что вдовствующие люди чаще возвращают долг без задержек, а не бывавшие в браке чаще задерживают оплату(гражданский - браком не считается). В середине списка мы видим имеющих опыт проживания в браке то есть состоящих в браке или разведенных. К тому же мы видим аналогичную зависимость связанную с возрастом, что значит:  
            **С возрастом и приобретенным опытом семейной жизни повышается личная ответственность индивида.**

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
data.groupby('total_income_category')['total_income_category'].count()

total_income_category
A       25
B     5014
C    15921
D      349
E       22
Name: total_income_category, dtype: int64

In [ ]:
display(data.groupby([data['total_income_category'] == 'A'])['purpose_category'])

In [ ]:
data.groupby('total_income_category')['debt'].sum()

total_income_category
A       2
B     354
C    1353
D      21
E       2
Name: debt, dtype: int64

In [ ]:
(data.groupby('total_income_category')['debt'].sum() / data.groupby('total_income_category')['total_income'].count() * 100)

total_income_category
A    8.000000
B    7.060231
C    8.498210
D    6.017192
E    9.090909
dtype: float64

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
**Вывод:**  


Исходя из проведенного исследования можно предположить что раз процентное соотношение невозврата кредита по группам с различным 
доходом невелико и нельзя выявить группу должников с большим отрывом от остальных исследуемых групп, то доход не является 
ключевым фактором влияющим на возврат кредита в срок, как и не является гарантом его возврата. Однако же, следует отметить что 
наименьший процент вернувших кредит в срок оказался именно в группе людей с наименьшим доходом. Отсюда следует простой
вывод - чем ниже у человека доход, тем больше риск невозврата кредита в срок. С другой стороны следующая группа должников с наибольшим процентом просрочки оказалась людьми с наибольшим доходом. Отсюда логично предположить что и суммы кредита, взятые ими, были значительно больше чем у предыдущей группы заемщиков. 

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
data.groupby('purpose_category')['purpose_category'].count().sort_values()

purpose_category
проведение свадьбы           2313
получение образования        3988
операции с автомобилем       4279
операции с недвижимостью    10751
Name: purpose_category, dtype: int64

In [ ]:
data.groupby('purpose_category')['debt'].sum().sort_values()

purpose_category
проведение свадьбы          183
получение образования       369
операции с автомобилем      400
операции с недвижимостью    780
Name: debt, dtype: int64

In [ ]:
(data.groupby('purpose_category')['debt'].sum() / data.groupby('purpose_category')['purpose_category'].count() * 100).sort_values()

purpose_category
операции с недвижимостью    7.255139
проведение свадьбы          7.911803
получение образования       9.252758
операции с автомобилем      9.347978
dtype: float64

**Вывод:** Проведенное исследование дает нам четко понять, что люди берущие кредит для свадьбы или операций с недвижимостью чаще возвращают его в срок, чем люди берущие кредит для получения образования или операций с автомобилем. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

Ниже проводим проверку для ответа на вопрос.

In [ ]:
data[data['days_employed'].isna()].count()

children                 0
days_employed            0
dob_years                0
education                0
education_id             0
family_status            0
family_status_id         0
gender                   0
income_type              0
debt                     0
total_income             0
purpose                  0
total_income_category    0
purpose_category         0
dtype: int64

In [ ]:
data[data['days_employed'].isna()]['income_type']

Series([], Name: income_type, dtype: object)

*Ответ:* Считаю что пропуски в столбце total_income возникли из-за пропусков в столбце days_employed, так алгоритм не мог посчитать значения в total_income без days_employed. Установить причину пропусков в столбце days_employed считаю невозможным в текущих условиях(надо спрашивать того кто выгружал или составлял DataFrame)

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Среднее значение может быть завышенным или заниженным из-за присутствия в выборке аномально больших или маленьких значений.

### Шаг 4: общий вывод.

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">

    
# Вывод

## Предисловие

1. В исходных данных было обнаруженно 2174 пропуска
2. Разброс чисел в категориях заемщиков при анализе оказался очень большим


Считаю эти 2 фактора могли повлиять на результаты исследования.  
Рекомендация:
1. Установить и исправить источник пропусков
2. Провести повторное исследование с большим количеством данных.

## Гипотезы

В результате проведенного исследования были проверены выдвинутые гипотезы:
   - Количество детей влияет на вероятность погашения кредита в срок (частично подтверждена)
   - Семейное положение влияет на вероятность погашения кредита в срок (подтверждена)
   - Уровень дохода влияет на вероятность погашения кредита в срок (требуется дополнительное исследование)
   - Цели кредита влияют на вероятность погашения кредита в срок (подтверждена)
    
В результате первая гипотеза подтвердилась лишь частично. Было установлено, что колличество детей не имеет значительного влияния на вероятность возврата кредита в срок, но **наличие детей или же их отсутствие имеет более значительное влияние.**

В ходе проверки второй гипозы была обнаружена прямая зависимость между семейным положением и возвратом кредита в срок, так был сделан вывод ,что вдовствующие люди чаще возвращают долг без задержек, а не бывавшие в браке чаще задерживают оплату. Средние позиции занимают люди имеющие опыт проживания в браке. К тому же мы видим аналогичную зависимость связанную с возрастом, что приводит нас к выводу:
**С возрастом и приобретенным опытом семейной жизни повышается личная ответственность индивида.**

Третья гипотеза не показала ожидаемых результатов, были выявлены похожие уровни просрочек как у людей с самым низким доходом, так и у людей с самым высоким уровнем дохода. В целом, разброс по категориям с разными уровнями доходы был сравнительно небольшим, что показывало малое влияние уровня дохода на вероятность просрочки или оплаты кредита вовремя.
**Следует провести повторное исследование на большем количестве данных.**

    
Четвертое исследование показало, что люди берущие кредит для свадьбы или операций с недвижимостью чаще возвращают его в срок, чем люди берущие кредит для получения образования или операций с автомобилем. 
**Гипотеза полностью подтверждена**

## Рекомендации

1. При принятии решения об одобрении выдачи кредита обращать повышенное внимание на:
- наличие/отсутствие детей
- возраст
- семейное положение
- цели кредита

2. Провести дополнительное исследование гипотезы "Уровень дохода влияет на вероятность погашения кредита в срок" на больших данных.


3. Выявить и устранить причины появления пропусков в столбцах о количестве проработаных дней и доходах.